Error handling: Malformed external address causes SIGSEGV

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): 0.22.6

**ABCI app** (name for built-in, URL for self-written if it's publicly available):  gaiad

**Environment**:
- **OS** (e.g. from /etc/os-release):  alpine 3.7
- **Install tools**:
- **Others**:


**What happened**:  Malformed external_address in gaiad's config.toml causes a segmentation fault (SIGSEGV) when attempting to write the node's own address to the addrbook.json


**What you expected to happen**:  Nice user friendly error message :)


**How to reproduce it** (as minimally and precisely as possible):
In config.toml set an invalid address (e.g. tcp://:26656 in this case). Run gaiad.

**Logs (you can paste a small part showing an error or link a pastebin, gist, etc. containing more of the log file)**:

```
I[07-25|22:27:27.809] Starting ABCI with Tendermint                module=main
I[07-25|22:27:27.835] Starting multiAppConn                        module=proxy impl=multiAppConn
I[07-25|22:27:27.836] Starting localClient                         module=abci-client connection=query impl=localClient
I[07-25|22:27:27.836] Starting localClient                         module=abci-client connection=mempool impl=localClient
I[07-25|22:27:27.836] Starting localClient                         module=abci-client connection=consensus impl=localClient
I[07-25|22:27:27.836] ABCI Handshake                               module=consensus appHeight=0 appHash=
I[07-25|22:27:27.836] ABCI Replay Blocks                           module=consensus appHeight=0 storeHeight=0 stateHeight=0
I[07-25|22:27:27.867] Completed ABCI Handshake - Tendermint and App are synced module=consensus appHeight=0 appHash=
I[07-25|22:27:27.867] This node is not a validator                 module=consensus addr=9A315A842B53D04FDA28C40D0B4A19C961A0ADE1 pubKey=PubKeyEd25519{F58E82C4BCED68E2C277307075158CEE22D5B1C60500F487251578D6841D7152}
I[07-25|22:27:27.880] Starting Node                                module=node impl=Node
I[07-25|22:27:27.880] Starting EventBus                            module=events impl=EventBus
I[07-25|22:27:27.880] Local listener                               module=p2p ip=:: port=26656
I[07-25|22:27:27.880] Starting DefaultListener                     module=p2p impl=Listener(@<nil>:26656)
I[07-25|22:27:27.880] P2P Node ID                                  module=node ID=4577de378e44fd6b009848d023e1aba04cbb8f75 file=/root/.gaiad/config/node_key.json
I[07-25|22:27:27.880] Add our address to book                      module=p2p book=/root/.gaiad/config/addrbook.json addr=null
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xb6ecc6]

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/p2p.(*NetAddress).String(0x0, 0xe8d08b, 0x17)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/p2p/netaddress.go:171 +0x26
github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/p2p/pex.(*addrBook).AddOurAddress(0xc4200de700, 0x0)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/p2p/pex/addrbook.go:164 +0x116
github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/node.(*Node).OnStart(0xc4200b59e0, 0xc420036900, 0xd)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/node/node.go:450 +0x4da
github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/libs/common.(*BaseService).Start(0xc4200b59e0, 0x1067260, 0xc4200fd170)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/libs/common/service.go:130 +0x3bd
github.com/cosmos/cosmos-sdk/server.startInProcess(0xc42089a2c0, 0xc42089a300, 0x1d, 0x0, 0x0)
	/go/src/github.com/cosmos/cosmos-sdk/server/start.go:111 +0x325
github.com/cosmos/cosmos-sdk/server.StartCmd.func1(0xc4208a2000, 0x17e23b8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cosmos/cosmos-sdk/server/start.go:38 +0xaa
github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra.(*Command).execute(0xc4208a2000, 0x17e23b8, 0x0, 0x0, 0xc4208a2000, 0x17e23b8)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra/command.go:698 +0x46d
github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc42086efc0, 0xd825a0, 0xc4200f3e01, 0xc42089a4c0)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra/command.go:783 +0x2e4
github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra.(*Command).Execute(0xc42086efc0, 0xc42089a4c0, 0xc4200f3ed8)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/libs/cli.Executor.Execute(0xc42086efc0, 0xfe7468, 0x2, 0xc420830e20)
	/go/src/github.com/cosmos/cosmos-sdk/vendor/github.com/tendermint/tendermint/libs/cli/setup.go:89 +0x4e
main.main()
	/go/src/github.com/cosmos/cosmos-sdk/cmd/gaia/cmd/gaiad/main.go:38 +0x214
```

**Config (you can paste only the changes you've made)**:

see above.

**`/dump_consensus_state` output for consensus bugs**

n/a 

**Anything else do we need to know**:

n/a

--- a/p2p/netaddress.go
+++ b/p2p/netaddress.go
@@ -13,6 +13,7 @@ import (
 	"strings"
 	"time"
 
+	"errors"
 	cmn "github.com/tendermint/tendermint/libs/common"
 )
 
@@ -97,16 +98,19 @@ func NewNetAddressStringWithOptionalID(addr string) (*NetAddress, error) {
 	if err != nil {
 		return nil, ErrNetAddressInvalid{addrWithoutProtocol, err}
 	}
+	if len(host) == 0 {
+		return nil, ErrNetAddressInvalid{
+			addrWithoutProtocol,
+			errors.New("host is empty")}
+	}
 
 	ip := net.ParseIP(host)
 	if ip == nil {
-		if len(host) > 0 {
-			ips, err := net.LookupIP(host)
-			if err != nil {
-				return nil, ErrNetAddressLookup{host, err}
-			}
-			ip = ips[0]
+		ips, err := net.LookupIP(host)
+		if err != nil {
+			return nil, ErrNetAddressLookup{host, err}
 		}
+		ip = ips[0]
 	}
 
 	port, err := strconv.ParseUint(portStr, 10, 16)

--- a/p2p/netaddress_test.go
+++ b/p2p/netaddress_test.go
@@ -49,6 +49,7 @@ func TestNewNetAddressStringWithOptionalID(t *testing.T) {
 
 		{"tcp://@127.0.0.1:8080", "", false},
 		{"tcp://@", "", false},
+		{"tcp://:26656", "", false},
 		{"", "", false},
 		{"@", "", false},
 		{" @", "", false},

